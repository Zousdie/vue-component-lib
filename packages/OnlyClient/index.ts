import Vue from 'vue';
import sfc from '../utils/sfc';

export interface IOnlyClientComponent {
  Slots: {
    default: ScopedSlot
    placeholder: ScopedSlot
  }
}

export interface OnlyClientParent extends Vue {
  $$isMounted?: boolean
}

export default sfc('only-client').def<IOnlyClientComponent>({
  functional: true,

  render(h, { parent, slots }) {
    const { default: slotsDefault, placeholder: slotsPlaceholder } = slots();

    if ((parent as OnlyClientParent).$$isMounted && slotsDefault) {
      return slotsDefault;
    }

    parent.$once('hook:mounted', () => {
      (parent as OnlyClientParent).$$isMounted = true;
      parent.$forceUpdate();
    });

    if (slotsPlaceholder) {
      return slotsPlaceholder;
    }

    return h();
  },
});
