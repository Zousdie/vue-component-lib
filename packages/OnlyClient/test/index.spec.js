import { mount } from '@vue/test-utils';
import OnlyClient from '..';

describe('Component OnlyClient', () => {
  it('render ssr', () => {
    const wrapper = mount({
      components: {
        OnlyClient,
      },
      template: `
        <div>
          <OnlyClient>
            only render on client
          </OnlyClient>
        </div>
      `,
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('render placeholder', () => {
    const wrapper = mount({
      components: {
        OnlyClient,
      },
      template: `
        <div>
          <OnlyClient>
            only render on client1
            only render on client2
            <template v-slot:placeholder>
              ssr placeholder
            </template>
          </OnlyClient>
        </div>
      `,
    });
    expect(wrapper.element).toMatchSnapshot();
  });
});
