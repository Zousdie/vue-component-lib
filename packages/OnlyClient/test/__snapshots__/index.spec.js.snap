// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component OnlyClient render placeholder 1`] = `
<div>
  
            only render on client1
            only render on client2
            
</div>
`;

exports[`Component OnlyClient render ssr 1`] = `
<div>
  <!---->
</div>
`;
